### ISSUE TITLE
Xiaomi MiFlora no battery level (HHCCJCY01)

### ISSUE THREAD

**Issue by kirbo** [custom-components/ble_monitor#405]
Hi,

I just switched from BT-MQTT-Gateway (https://github.com/zewelor/bt-mqtt-gateway) into using ESP32 with HHCCJCY01 platform.
I noticed that in the Readme is said:
> MiFlora plant sensor
>
> Broadcasts temperature, moisture, illuminance, conductivity, 1 reading per minute, no battery info with firmware v3.2.1.

...but I'd like to question is this still the case?
By using the bt-mqtt-gateway, I was able to retrieve even the battery level, even though all of my sensors are on firmware v3.3.1, as well as I can see the battery level via phone, using either VegTrug or Flower Care apps on iPhone.

So, basically this ticket is about to question why is other projects/softwares able to retrieve the battery level, but not this one?

**Comment by Ernst79** [comment]
I will check it, I have the device myself. Maybe they added it in the latest firmware, but I have to check. A have a bit trouble finding the firmware version in MiHome, as 70% of the text has turned to Chinese.

**Comment by Ernst79** [comment]
Anyway, let me explain why we can't get the battery state. BT-MQTT-gateway is using Miflora to get the data. This package is setting up a connection to the plant sensor, and "asks" for the battery state. BLE monitor is a passive component, just listening to BLE advertisements that are send by the sensor (without asking for it). The advantage of this is that it isn't affecting the battery of the sensors, the downside it that is doesn't receive the battery state. HHCCJCY01 seems to be the one of the few sensors that isn't sending the battery state out of its own. No idea why they have decided so. 

there is a workaround. 

Just ask for the battery state once a day with the official component. This will save your battery and you can use it together with BLE monitor for the other data.  

```
sensor:
  - platform: miflora
      mac: 'XX:XX:XX:XX:XX:XX'
      name: miflora
      force_update: false
      scan_interval: 3600 #battery level every hour
      monitored_conditions:
        - battery
```

**Comment by kirbo** [comment]
Oh, okay, I see. I didn't recognise that this is just listening the BLE messages and BT-MQTT-gateway was polling..

Should this work with ESPHome? I tried adding the sensor you mentioned above and I get error `Platform not found: 'sensor.miflora'.`

**Comment by kirbo** [comment]
Anywho, closing this issue, since there seems to be others related to this very same issue 🙂
Thanks anyway!

**Comment by Ernst79** [comment]
Probably not, this is the BLE monitor repo 😄 . 

I think ESPHome is also passively listening, but you can also do polling since one of the recent updates. Anyway, you could better ask on the ESPhome repo

**Comment by mrmaximas** [comment]
@Kirbo @Ernst79 have you solved the problem of polling the battery status of HHCCJCY01?

**Comment by Ernst79** [comment]
You can use an ESPHome directly for miflora devices

https://esphome.io/components/sensor/xiaomi_hhccjcy10.html

**Comment by mrmaximas** [comment]
i can try to poll via this config but i don't know service_uuid characteristic_uuid or this is no right way?

esp32_ble_tracker:

ble_client:
  - mac_address: C4:7C:8D:6A:3C:A1
   id: test

sensor:

  - platform: ble_client
    type: characteristic
    ble_client_id: test
   name: "Test Battery"
  service_uuid: ?
   characteristic_uuid: ?
  update_interval: 10min

**Comment by Ernst79** [comment]
The 01 is also available for ESPHome, but check the note

**Comment by mrmaximas** [comment]
@Ernst79 Above I wrote that I can try `ble_client` but I don’t know how to get and parse the response.

**Comment by Ernst79** [comment]
Characteristic uuid is "00001a02-0000-1000-8000-00805f9b34fb"

Service uuid is most likely "0000fe95-0000-1000-8000-00805f9b34fb"

Got it from the xiaomi-ble integration

https://github.com/Bluetooth-Devices/xiaomi-ble/blob/main/src/xiaomi_ble/const.py

**Comment by mrmaximas** [comment]
@Ernst79 [17:55:17][W][ble_sensor:051]: No sensor characteristic found at service 00001A02-0000-1000-8000-00805F9B34FB char 0000FE95-0000-1000-8000-00805F9B34FB

**Comment by Ernst79** [comment]
looks like you switched the service UUid and Characteristic UUid

**Comment by mrmaximas** [comment]
@Ernst79 
[18:18:59][D][esp32_ble_client:418]: [0] [C4:7C:8D:6D:4C:AF] Event 46
[18:18:59][D][esp32_ble_client:193]: [0] [C4:7C:8D:6D:4C:AF] ESP_GATTC_SEARCH_CMPL_EVT
[18:18:59][I][esp32_ble_client:339]: [0] [C4:7C:8D:6D:4C:AF] Connected
[18:18:59][D][sensor:098]: 'Test Battery': Sending state nan  with 0 decimals of accuracy
[18:18:59][W][ble_sensor:051]: No sensor characteristic found at service 0000FE95-0000-1000-8000-00805F9B34FB char 00001A02-0000-1000-8000-00805F9B34FB

**Comment by Ernst79** [comment]
than you need to figure out what to use with these instructions
https://esphome.io/components/ble_client.html#setting-up-devices

**Comment by Ernst79** [comment]
Can you try to add with the auto connect option?

esp32_ble_tracker:

ble_client:
  - mac_address: XX:XX:XX:XX:XX:XX
    id: itag_black
    auto_connect: true

**Comment by mrmaximas** [comment]
@Ernst79 hi! this code works for me https://github.com/myhomeiot/esphome-components/blob/main/examples/myhomeiot_ble_client/flower_care_battery_ble_advertise.yaml
```
[09:07:59][D][esp-idf:000][BTU_TASK]: W (683282) BT_HCI: hci cmd send: disconnect: hdl 0x0, rsn:0x13
[09:07:59][D][esp-idf:000][BTU_TASK]: W (683374) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x16
[09:09:00][D][myhomeiot_ble_client:069]: [80:EA:CA:89:20:71] Found device
[09:09:00][D][esp32_ble_tracker:234]: Stopping scan to make connection
[09:09:00][I][myhomeiot_ble_client:036]: [80:EA:CA:89:20:71] Connecting
[09:09:00][D][esp32_ble_tracker:116]: connecting: 1, discovered: 0, searching: 0, disconnecting: 0
[09:09:00][D][esp32_ble_tracker:343]: End of scan, set scanner state to IDLE.
[09:09:01][I][myhomeiot_ble_client:087]: [80:EA:CA:89:20:71] Connected successfully, app_id (1)
[09:09:01][D][esp32_ble_tracker:116]: connecting: 0, discovered: 0, searching: 0, disconnecting: 0
[09:09:01][D][esp32_ble_tracker:307]: Starting scan, set scanner state to STARTING.
[09:09:07][D][myhomeiot_ble_client:129]: [80:EA:CA:89:20:71] SEARCH_RES_EVT service (0x1204) found
[09:09:07][D][myhomeiot_ble_client:166]: [80:EA:CA:89:20:71] SEARCH_CMPL_EVT char (0x1A02) found
[09:09:07][I][myhomeiot_ble_client:067]: Battery (100%), firmware (3.3.5)
[09:09:07][I][myhomeiot_ble_client:046]: [80:EA:CA:89:20:71] Disconnecting
[09:09:07][D][esp-idf:000][BTU_TASK]: W (751562) BT_HCI: hci cmd send: disconnect: hdl 0x0, rsn:0x13
[09:09:07][D][esp-idf:000][BTU_TASK]: W (751708) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x16
[09:09:21][D][myhomeiot_ble_client:069]: [C4:7C:8D:6E:05:A5] Found device
[09:09:21][D][esp32_ble_tracker:234]: Stopping scan to make connection
[09:09:21][I][myhomeiot_ble_client:036]: [C4:7C:8D:6E:05:A5] Connecting
[09:09:21][D][esp32_ble_tracker:116]: connecting: 1, discovered: 0, searching: 0, disconnecting: 0
[09:09:21][D][esp32_ble_tracker:343]: End of scan, set scanner state to IDLE.
[09:09:27][I][myhomeiot_ble_client:087]: [C4:7C:8D:6E:05:A5] Connected successfully, app_id (1)
[09:09:27][D][esp32_ble_tracker:116]: connecting: 0, discovered: 0, searching: 0, disconnecting: 0
[09:09:27][D][esp32_ble_tracker:307]: Starting scan, set scanner state to STARTING.
[09:09:33][D][myhomeiot_ble_client:129]: [C4:7C:8D:6E:05:A5] SEARCH_RES_EVT service (0x1204) found
[09:09:33][D][myhomeiot_ble_client:166]: [C4:7C:8D:6E:05:A5] SEARCH_CMPL_EVT char (0x1A02) found
[09:09:34][I][myhomeiot_ble_client:111]: Battery (42%), firmware (3.3.5)
[09:09:34][I][myhomeiot_ble_client:046]: [C4:7C:8D:6E:05:A5] Disconnecting
[09:09:34][D][esp-idf:000][BTU_TASK]: W (778095) BT_HCI: hci cmd send: disconnect: hdl 0x0, rsn:0x13
[09:09:34][D][esp-idf:000][BTU_TASK]: W (778177) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x16
[09:09:43][D][myhomeiot_ble_client:069]: [5C:85:7E:B1:01:84] Found device
[09:09:43][D][esp32_ble_tracker:234]: Stopping scan to make connection
[09:09:43][I][myhomeiot_ble_client:036]: [5C:85:7E:B1:01:84] Connecting
[09:09:43][D][esp32_ble_tracker:116]: connecting: 1, discovered: 0, searching: 0, disconnecting: 0
[09:09:43][D][esp32_ble_tracker:343]: End of scan, set scanner state to IDLE.
[09:09:46][I][myhomeiot_ble_client:087]: [5C:85:7E:B1:01:84] Connected successfully, app_id (1)
[09:09:46][D][esp32_ble_tracker:116]: connecting: 0, discovered: 0, searching: 0, disconnecting: 0
[09:09:46][D][esp32_ble_tracker:307]: Starting scan, set scanner state to STARTING.
[09:09:52][D][myhomeiot_ble_client:129]: [5C:85:7E:B1:01:84] SEARCH_RES_EVT service (0x1204) found
[09:09:52][D][myhomeiot_ble_client:166]: [5C:85:7E:B1:01:84] SEARCH_CMPL_EVT char (0x1A02) found
[09:09:52][I][myhomeiot_ble_client:089]: Battery (0%), firmware (3.3.6)
[09:09:52][I][myhomeiot_ble_client:046]: [5C:85:7E:B1:01:84] Disconnecting
[09:09:52][D][esp-idf:000][BTU_TASK]: W (796750) BT_HCI: hci cmd send: disconnect: hdl 0x0, rsn:0x13
[09:09:52][D][esp-idf:000][BTU_TASK]: W (796844) BT_HCI: hcif disc complete: hdl 0x0, rsn 0x16
```
But I don't use Home Assistant at all. I collect data to mqtt broker. Please help me send MiFlora battery data to mqtt.
Thanks!
